h2/src/main/org/h2/engine/Session.java

```diff
@@ -1814,39 +1814,38 @@ private long getStatementSavepoint() {
     public void startStatementWithinTransaction(Command command) {
         Transaction transaction = getTransaction();
         if (transaction != null) {
-            HashSet<MVMap<?, ?>> currentMaps = null, allMaps = null;
+            HashSet<MVMap<?, ?>> maps = new HashSet<>();
             if (command != null) {
                 Set<DbObject> dependencies = command.getDependencies();
-                currentMaps = new HashSet<>();
-                for (DbObject dependency : dependencies) {
-                    if (dependency instanceof MVTable) {
-                        addTableToDependencies((MVTable) dependency, currentMaps);
-                    }
-                }
                 switch (transaction.getIsolationLevel()) {
-                case REPEATABLE_READ: {
-                    allMaps = new HashSet<>();
-                    HashSet<MVTable> processed = new HashSet<>();
-                    for (DbObject dependency : dependencies) {
-                        if (dependency instanceof MVTable) {
-                            addTableToDependencies((MVTable) dependency, allMaps, processed);
-                        }
-                    }
-                    break;
-                }
                 case SNAPSHOT:
                 case SERIALIZABLE:
                     if (!transaction.hasStatementDependencies()) {
-                        allMaps = new HashSet<>();
                         for (Table table : database.getAllTablesAndViews(false)) {
                             if (table instanceof MVTable) {
-                                addTableToDependencies((MVTable) table, allMaps);
+                                addTableToDependencies((MVTable)table, maps);
                             }
                         }
                     }
+                    //$FALL-THROUGH$
+                case REPEATABLE_READ:
+                    HashSet<MVTable> processed = new HashSet<>();
+                    for (DbObject dependency : dependencies) {
+                        if (dependency instanceof MVTable) {
+                            addTableToDependencies((MVTable)dependency, maps, processed);
+                        }
+                    }
+                    break;
+                case READ_COMMITTED:
```

Comment:
I think that `READ UNCOMMITTED` needs snapshot at start of the command too. The only differente between this level and default `READ COMMITTED` is visibility of uncommitted changes from other transactions, other differences aren't expected.

We had a bug that you fixed some time ago with consistency of statements. The same bug may re-appear here with `READ UNCOMMITTED` isolation level. H2 has an optimization for `MIN` / `MAX` / `COUNT(*)` aggregates. They are evaluated separately (and they can use different indexes / maps). But even on this weak level these indexes / maps must be consistent with each other during evaluation of the query to retain the same behavior as with unoptimized version.

Reply:
My understanding of READ UNCOMMITTED is that it always will read latest changes possible, with minimal overhead, and consistency is not guaranteed. All those aggregates will be broken in face of concurrent transactions anyway. Yes, there could be subtle inconsistencies, even if transaction runs alone, but why would you choose this isolation level, in a first place?

Reply:
During the normal execution of a query all rows are processed one by one. This weak level may get a mixed set of rows, but each row either visible or not, it cannot be partially visible. During optimized execution of a grouped query H2 can avoid table scan at all and perform some lookups in indexes. When we perform such optimization its results should be the same. If a table had rows (1, 1), (2, 2) and a row (3, 3) was inserted during execution of `SELECT MAX(C1), MAX(C2), COUNT(*) FROM tableName`, the result may be (2, 2, 2), or (3, 3, 3), but it shouldn't be (2, 2, 3) or something like it.

Some other databases perform such optimization only when there is only one aggregate function, possibly due to this problem with consistency. H2 performs such optimization unconditionally with any number of aggregates.

Reply:
The situation could be even worse if there are primary and secondary keys with the same sorting position. (1, '000001'), (2, '000002'), for example. `SELECT MAX(ID), MAX(CODE)` shouldn't return (1, '000002').